### PR TITLE
Add indexes for train, test and validation.

### LINKED ISSUE #21: Train / test splits
Several repositories, such as the UCI or UCR, provide different datasets for train and test. Currently, we store these as separate attributes inside the Bunch object. This has several downsides:

* This approach mixes the data and the task (consider as an example OpenML where these concepts are well separated).
* Sometimes (most of the time?) the user is interested in the whole dataset, and not in the original task. Having to select and concatenate both attributes is confusing.
* If we want to support a `return_X_y` parameter this setup is confusing. Should we return only train, or a 4-tuple, breaking compatibility with scikit-learn?
* If we want to provide `as_frame`, would we need `frame` and `frame_test` attributes?
* If we want to provide `as_frame`, the names of the columns would be repeated in both dataframes (not very important, but still).

As an alternative, we could consider putting all data together and providing an attribute(s) with the indexes/size of the train/test data. Extracting the train/test sets would be trivial, if desired, and the normal behaviour would be to provide all the data as one attribute.

### PR DESCRIPTION
**Pull request by vnmabus**
Fixes #21.

I doubt that the current tests are enough to cover everything.

We must discuss:
- The naming of the new Bunch attributes.
- Whether to change the way in which we provide cross validators.
- If it is better to split train and validation indexes or not (or both).
- If this is going to be a breaking change (as it is now) or some deprecation scheme must be devised.

### CODE REVIEW

**Comment by daviddiazvico** [comment]
Looks good.

So now, if I have a ``Bunch`` named ``data`` that includes a train/test partition, I just need to do the following to extract the different partitions and whole dataset:

    X_train = data.data[data.train_indexes]
    y_train = data.target[data.train_indexes]
    X_test = data.data[data.test_indexes]
    y_test = data.target[data.test_indexes]
    X = data.data
    y = data.target

This looks very intuitive, I like it.

**Comment by vnmabus** [comment]
Yes, that is right. The only problem is if you have a validation set and you want to pick the train and validation together, as this is often the case. In that case you would need to check if there is a validation set and use `np.concatenate`.

I am thinking about this problem, and now I think that is better that the indexes are lists instead of ndarrays, so that you can concatenate them using +. We can also replace `None` by the empty list, so you still could check absence with `if data.validation_indexes:` but you could concatenate without checking. If we apply these changes, then in order to pick train and validation together you would do
```python
X_train = data.data[data.train_indexes + data.validation_indexes]
```
and it will work even if the dataset does not have a validation set.

**Comment by daviddiazvico** [comment]
Nice. I agree that a list is a better choice for the indices than a np.array. Please ping me if you want me to review any code. Thanks.

**Comment by vnmabus** [comment]
The tests are failing because if there is no division in tests, `train_indexes` will be `[]`, as will be the other indexes. Another option would be to consider ALL data as the train set, but I am not comfortable with that option, even when the existing tests assume it. @daviddiazvico, what do you think?

**Comment by daviddiazvico** [comment]
I don't like the idea of considering all data as the train set. Maybe I'm not understanding you, but why not just changing the tests to check each of the partitions separately?

**Comment by vnmabus** [comment]
I was talking about the case where there are no predefined partitions. Should the data be considered as one (train) partition, or we consider that the partitions are empty, and let the user deal with that?

**Comment by daviddiazvico** [comment]
Ah, ok. I prefer to let the user decide in that case. I understand that when no partitions are available, then the correct approach is CV, but don't want to force that.

**Comment by vnmabus** [comment]
Then I will change the tests accordingly.

**Comment by vnmabus** [comment]
Sorry for the delay in fixing this. I think we only need to discuss the points in the PR description before accepting this.

**Comment by daviddiazvico** [comment]
We must discuss:

- The naming of the new Bunch attributes. -> **train_indexes, validation_indexes and test_indexes are ok for me (which is preferred, "indexes" or "indices"?), the rest remain the same, am I right?**
- Whether to change the way in which we provide cross validators. -> **what do you mean exactly? I understand you have not changed this for now.**
- If it is better to split train and validation indexes or not (or both). -> **well, whatever the source of the dataset specifies, if I understand correctly.**
- If this is going to be a breaking change (as it is now) or some deprecation scheme must be devised. -> **I'm ok with a breaking change. Older versions will remain available at Github and Pypi.**

**Comment by vnmabus** [comment]
>     * The naming of the new Bunch attributes. -> **train_indexes, validation_indexes and test_indexes are ok for me (which is preferred, "indexes" or "indices"?), the rest remain the same, am I right?**

Both are accepted, I think. It looks to me that `numpy` uses "indices" more often, so I will rename these to match.

>     * Whether to change the way in which we provide cross validators. -> **what do you mean exactly? I understand you have not changed this for now.**

I haven't. But now the `inner_cv` attribute is somehow redundant, as it can be done outside the library (or with a helper method) given the train and cv indexes. Also the `outer_cv` could probably be defined using a set of indexes too.

>     * If it is better to split train and validation indexes or not (or both). -> **well, whatever the source of the dataset specifies, if I understand correctly.**

True. This has been already answered.

>     * If this is going to be a breaking change (as it is now) or some deprecation scheme must be devised. -> **I'm ok with a breaking change. Older versions will remain available at Github and Pypi.**

Ok.

**Comment by vnmabus** [comment]
I have changed indexes to indices. It seems that the tests are failing because the API for forex is down (https://github.com/MicroPyramid/forex-python/issues/86). As the python_forex repo seems unmaintained, probably we should move to a different dependency in another PR.

As the remaining tests pass, I will open this PR for merging. If we want to change the CV API that can be done in a different PR.
